HW2/P2/P2.txt

```diff
@@ -0,0 +1,35 @@
+1. When serial uncorrelated
+serial uncorrelated: 0.389642000198 seconds
+Fine grained uncorrelated: 5.39227604866 seconds
+
+Medium grained uncorrelated:
+N=5, 7.78783583641 seconds
+N=10, 8.25664401054 seconds
+N=15, 8.96474099159 seconds
+N=20, 9.40219688416 seconds
+N=50, 10.1232030392 seconds
+
+
+2. When serial correlated
+Serial correlated: 0.448005914688 seconds
+Fine grained correlated: 4.11850094795 seconds
+
+Medium grained correlated:
+N=5, 6.68246507645 seconds
+N=10, 6.85533285141 seconds
+N=15, 7.27448797226 seconds
+N=20, 8.3435189724 seconds
+N=50, 9.59883904457 seconds
+
```

Comment:
I saw the same relationship between the three approaches, i.e. serial runtime << fine grained runtime < medium-grained runtime.
I used a different locking strategy, where I locked both the source and destination to avoid any value going < 0 (see my comment below). My uncorrelated medium-grained graph looks roughly similar, but my correlated medium-grained graph has a different shape due to my locking strategy - i.e. it bottoms out at ~N=20.
I'm guessing that the "correct" approach depends on how important that <0 condition is, as your approach does ensure that multiple threads can't update the same value concurrently.